### PR TITLE
Dnstap plugin refactoring

### PR DESCRIPTION
**Pull request by UladzimirTrehubenka**
### 1. What does this pull request do?
improved initializaton (start/restart/shutdown);
avoid using buffer for golang-framestream encoder (it is bufio);
avoid make TCP connection (reuse one time created TCP connection) for each Write() call;
reuse code for Socket and TCP connection;
use Flush() call each 1 sec;
use chan queue (1000 entries) for avoid golang-framestream encoder race;
speed up dnstap plugin performance.
### 2. Which issues (if any) are related?
none
### 3. Which documentation changes (if any) need to be made?
none

### CODE REVIEW

File: plugin/dnstap/dnstapio/io_test.go [hunk 1]
```diff
@@ -1,78 +1,109 @@
...
-					wg.Done()
-					return
-				default:
-					time.Sleep(50 * time.Millisecond)
-					t := tap.Dnstap_MESSAGE
-					dio.Dnstap(tap.Dnstap{Type: &t})
-				}
-			}
-		}()
+const endpointTCP = "localhost:14000"
```
Comment (miekg) [inline plugin/dnstap/dnstapio/io_test.go]: This will probably be problematic in testing; is it possible to listen on localhost:0?

File: plugin/dnstap/dnstapio/io.go [hunk 2]
```diff
@@ -2,67 +2,99 @@ package dnstapio
...
 
-func (dio *DnstapIO) serve() {
+// Close waits until the I/O routine is finished to return.
+func (dio *dnstapIO) Close() {
+	close(dio.queue)
+	dio.conn.Close()
+}
+
+func (dio *dnstapIO) serve() {
+	timeout := time.After(flushTimeout)
```
Comment (miekg) [inline plugin/dnstap/dnstapio/io.go]: make this a time.Ticker? Resolves the need to reassign to timeout ?

File: plugin/dnstap/dnstapio/io.go [hunk 3]
```diff
@@ -2,67 +2,99 @@ package dnstapio
...
+	Close()
 }
 
-// New dnstap I/O routine from Protocol.
-func New(w Protocol) *DnstapIO {
-	dio := DnstapIO{}
-	dio.protocol = w
-	dio.queue = make(chan tap.Dnstap, 10)
-	dio.stop = make(chan bool)
+// Connect to the Dnstap endpoint
```
Comment (miekg) [inline plugin/dnstap/dnstapio/io.go]: Connect connects to the dnstop endpoint.

File: plugin/dnstap/dnstapio/io.go [hunk 4]
```diff
@@ -2,67 +2,99 @@ package dnstapio
...
+	enc   *fs.Encoder
+	conn  net.Conn
+	queue chan tap.Dnstap
 }
 
-// Protocol is either `out.TCP` or `out.Socket`.
-type Protocol interface {
-	// Write takes a single frame at once.
-	Write([]byte) (int, error)
+// New dnstap I/O routine.
```
Comment (miekg) [inline plugin/dnstap/dnstapio/io.go]: New returns a new and initialized DnstapIO

File: plugin/dnstap/dnstapio/io.go [hunk 5]
```diff
@@ -2,67 +2,99 @@ package dnstapio
 
 import (
 	"log"
+	"net"
+	"time"
 
 	tap "github.com/dnstap/golang-dnstap"
+	fs "github.com/farsightsec/golang-framestream"
```
Comment (miekg) [inline plugin/dnstap/dnstapio/io.go:9]: We need a dep ensure for this (but should do this in another PR)

**Review by miekg — COMMENTED** [review]
Nice cleanup!

File: plugin/dnstap/dnstapio/io.go [hunk 6]
```diff
@@ -2,67 +2,99 @@ package dnstapio
...
+}
+
+func (dio *dnstapIO) serve() {
+	timeout := time.After(flushTimeout)
 	for {
 		select {
-		case payload := <-dio.queue:
+		case payload, ok := <-dio.queue:
+			if !ok {
+				return
```
Comment (rdrozhdzh) [inline plugin/dnstap/dnstapio/io.go]: shouldn't we flush here?

File: plugin/dnstap/dnstapio/io.go (hunk 4 above)
Comment (UladzimirTrehubenka) [reply plugin/dnstap/dnstapio/io.go]: Fixed

File: plugin/dnstap/dnstapio/io.go (hunk 3 above)
Comment (UladzimirTrehubenka) [reply plugin/dnstap/dnstapio/io.go]: Fixed

File: plugin/dnstap/dnstapio/io_test.go (hunk 1 above)
Comment (UladzimirTrehubenka) [reply plugin/dnstap/dnstapio/io_test.go]: Fixed

File: plugin/dnstap/dnstapio/io.go (hunk 2 above)
Comment (UladzimirTrehubenka) [reply plugin/dnstap/dnstapio/io.go]: There is a concern - if Flush() takes more time than ticker interval - we have useless pending Flush() call(s).

File: plugin/dnstap/dnstapio/io.go [hunk 7]
```diff
@@ -2,67 +2,99 @@ package dnstapio
...
 			frame, err := proto.Marshal(&payload)
-			if err == nil {
-				dio.protocol.Write(frame)
-			} else {
-				log.Printf("[ERROR] Invalid dnstap payload dropped: %s\n", err)
+			if err != nil {
+				log.Printf("[ERROR] Invalid dnstap payload dropped: %s", err)
+				continue
+			}
+			_, err = dio.enc.Write(frame)
```
Comment (rdrozhdzh) [inline plugin/dnstap/dnstapio/io.go:88]: Can it reconnect if connection was lost?

File: plugin/dnstap/dnstapio/io_test.go [hunk 8]
```diff
@@ -1,78 +1,109 @@
...
+	}
 
-func (b buf) Close() error {
-	return nil
+	if err := server.Close(); err != nil {
+		t.Error(err)
+	}
 }
 
-func TestRace(t *testing.T) {
```
Comment (rdrozhdzh) [inline plugin/dnstap/dnstapio/io_test.go:32]: I think the concurrent test is important, we need to preserve it

File: plugin/dnstap/dnstapio/io.go (hunk 7 above)
Comment (UladzimirTrehubenka) [reply plugin/dnstap/dnstapio/io.go:88]: No

File: plugin/dnstap/dnstapio/io.go (hunk 6 above)
Comment (UladzimirTrehubenka) [reply plugin/dnstap/dnstapio/io.go]: Fixed

File: plugin/dnstap/dnstapio/io_test.go (hunk 8 above)
Comment (UladzimirTrehubenka) [reply plugin/dnstap/dnstapio/io_test.go:32]: Fixed
